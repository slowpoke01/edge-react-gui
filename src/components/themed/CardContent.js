// @flow

import * as React from 'react'
import { View } from 'react-native'

import { unpackEdges } from '../../util/edges'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { DataRow } from './DataRow'
import { EdgeText } from './EdgeText.js'

type Props = {
  image?: React.Node,
  title: string | React.Node,
  subTitle: string,
  value: string | React.Node,
  subValue: string,
  paddingRem?: number[] | number
}

export class CardContentComponent extends React.PureComponent<Props & ThemeProps> {
  renderTitle() {
    const { title, theme } = this.props
    const styles = getStyles(theme)
    if (typeof title === 'string') return <EdgeText style={styles.contentTitle}>{title}</EdgeText>

    return title
  }

  renderValue() {
    const { value, theme } = this.props
    const styles = getStyles(theme)

    if (typeof value === 'string') return <EdgeText style={styles.contentValue}>{value}</EdgeText>

    return value
  }

  render() {
    const { image, subTitle, subValue, paddingRem, theme } = this.props
    const styles = getStyles(theme)

    return (
      <View style={[styles.container, paddingStyles(paddingRem, theme)]}>
        {image ? <View style={styles.iconContainer}>{image}</View> : null}
        <View style={styles.contentContainer}>
          <DataRow label={this.renderTitle()} value={this.renderValue()} />
          <DataRow
            label={<EdgeText style={styles.contentSubTitle}>{subTitle}</EdgeText>}
            value={<EdgeText style={styles.contentSubValue}>{subValue}</EdgeText>}
          />
        </View>
      </View>
    )
  }
}

function paddingStyles(paddingRem?: number[] | number, theme: Theme) {
  const padding = unpackEdges(paddingRem == null ? 0 : paddingRem)

  return {
    paddingBottom: theme.rem(padding.bottom),
    paddingLeft: theme.rem(padding.left),
    paddingRight: theme.rem(padding.right),
    paddingTop: theme.rem(padding.top)
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    flex: 1,
    flexDirection: 'row'
  },
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    marginRight: theme.rem(1)
  },
  contentContainer: {
    flex: 1,
    flexDirection: 'column'
  },
  contentTitle: {
    fontFamily: theme.fontFaceMedium,
    fontWeight: '600'
  },
  contentValue: {
    fontWeight: '600',
    textAlign: 'right'
  },
  contentSubTitle: {
    flex: 1,
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  contentSubValue: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  }
}))

export const CardContent = withTheme(CardContentComponent)
