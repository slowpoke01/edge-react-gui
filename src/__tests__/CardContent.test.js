/* globals describe it expect */
/* eslint-disable flowtype/require-valid-file-annotation */

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { getTheme } from '../components/services/ThemeContext.js'
import { CardContentComponent as Request } from '../components/themed/CardContent.js'

describe('Request', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      image: 11,
      title: 'string' | 11,
      subTitle: 'string',
      value: 'string' | 11,
      subValue: 'string',
      paddingRem: [11] | 11,
      theme: getTheme()
    }
    const actual = renderer.render(<Request {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
