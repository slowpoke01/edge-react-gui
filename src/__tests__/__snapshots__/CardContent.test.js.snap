// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request should render with loading props 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
      Object {
        "paddingBottom": 247,
        "paddingLeft": 247,
        "paddingRight": 247,
        "paddingTop": 247,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginRight": 22,
      }
    }
  >
    11
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <WithTheme(DataRowComponent)
      label={11}
      value={11}
    />
    <WithTheme(DataRowComponent)
      label={
        <WithTheme(EdgeTextComponent)
          style={
            Object {
              "color": "#A4C7DF",
              "flex": 1,
              "fontSize": 17,
            }
          }
        >
          string
        </WithTheme(EdgeTextComponent)>
      }
      value={
        <WithTheme(EdgeTextComponent)
          style={
            Object {
              "color": "#A4C7DF",
              "fontSize": 17,
            }
          }
        >
          string
        </WithTheme(EdgeTextComponent)>
      }
    />
  </View>
</View>
`;
